### PR TITLE
(PUP-10318) Fix errors when running multithreaded with environment cache on

### PR DESCRIPTION
**Pull request by pcarlisle**
Some of this is hacky for now, working on cleaning it up but I'm putting this up as a draft so it can be used for testing.

### CODE REVIEW

File: lib/puppet/pops/loaders.rb [hunk 1]
```diff
@@ -19,12 +19,14 @@ class LoaderError < Puppet::Error; end
   attr_reader :environment
 
   def self.new(environment, for_agent = false)
-    obj = environment.loaders
-    if obj.nil?
-      obj = self.allocate
-      obj.send(:initialize, environment, for_agent)
+    environment.lock.synchronize do
```
Comment (justinstoller) [inline lib/puppet/pops/loaders.rb]: Did you ever find out if this is an extension point external consumers (or even Bold) could use?

File: lib/puppet/type.rb [hunk 2]
```diff
@@ -84,6 +85,10 @@ class Type
   # Comparing type instances.
   include Comparable
 
+  @types = {}
+  # Lock for managing types
+  @lock = Puppet::Concurrent::Lock.new
```
Comment (justinstoller) [inline lib/puppet/type.rb]: This is nuts. Would it at least make sense to move this closer to where we include MetatypeManager, and maybe comment `@types`?

Comment (pcarlisle) [reply lib/puppet/type.rb]: Yeah that makes sense at least. I hate this a lot.

**Comment by justinstoller** [comment]
jenkins please test this with servertests

**Comment by mwaggett** [comment]
I don't really understand this because I don't have a ton of puppet OR threading experience, but this seems sane to me.. ¯\_(ツ)_/¯

**Pull request merged**